enterprise/server/sociartifactstore/sociartifactstore.go

```diff
@@ -285,112 +344,191 @@ func deserializeDigest(s string) (*repb.Digest, error) {
-				Hash:      strings.ReplaceAll(entry.Digest, "sha256:", ""),
-				SizeBytes: entry.Size,
-			}
-			return nil
-		}
-		return nil
-	})
+	digest, err := image.Digest()
 	if err != nil {
 		return nil, err
 	}
-	return sociIndex, nil
-}
-
-func toDigest(hash string) (*repb.Digest, error) {
-	stat, err := os.Stat(blobPath(hash))
+	size, err := image.Size()
 	if err != nil {
 		return nil, err
 	}
-	return &repb.Digest{
-		Hash:      hash,
-		SizeBytes: stat.Size(),
+	return &ocispec.Descriptor{
+		MediaType: string(mediaType),
+		Digest:    godigest.Digest(digest.String()),
+		Size:      size,
 	}, nil
 }
 
-func (s *SociArtifactStore) writeArtifactsToCache(ctx context.Context, imageId string, sociIndexDigest *repb.Digest, ztocDigests []*repb.Digest) error {
-	if _, err := s.blobstore.WriteBlob(ctx, blobKey(imageId), []byte(serializeDigest(sociIndexDigest))); err != nil {
-		return err
+func (s *SociArtifactStore) indexLayer(ctx context.Context, layer v1.Layer) (*repb.Digest, error) {
+	mediaType, err := layer.MediaType()
+	if err != nil {
+		return nil, err
 	}
-	if err := s.writeArtifactToCache(ctx, sociIndexDigest); err != nil {
-		return err
+	layerDigest, err := layer.Digest()
+	if err != nil {
+		return nil, err
 	}
-	for _, ztocDigest := range ztocDigests {
-		if err := s.writeArtifactToCache(ctx, ztocDigest); err != nil {
-			return err
-		}
+	layerSize, err := layer.Size()
+	if err != nil {
+		return nil, err
 	}
-	return nil
-}
 
-func (s *SociArtifactStore) writeArtifactToCache(ctx context.Context, digest *repb.Digest) error {
-	bytes, err := os.ReadFile(blobPath(digest.Hash))
+	compressionAlgo, err := images.DiffCompression(ctx, string(mediaType))
 	if err != nil {
-		return err
+		return nil, status.NotFoundErrorf("could not determine layer compression: %s", err)
+	}
+	if compressionAlgo != compression.Gzip {
+		return nil, status.UnimplementedErrorf("layer %s (%s) cannot be indexed because it is compressed with %s",
+			layerDigest.Hex, mediaType, compressionAlgo)
+	}
+
+	layerTmpFile, err := os.CreateTemp("", "layer.*")
+	if err != nil {
+		return nil, err
+	}
+	layerReader, err := layer.Compressed()
+	if err != nil {
+		return nil, err
+	}
+	numBytes, err := io.Copy(layerTmpFile, layerReader)
+	if err != nil {
+		return nil, err
+	}
+	if numBytes != layerSize {
+		return nil, status.DataLossErrorf("written layer size does not match that of the digest")
+	}
+
+	ztocBuilder := ztoc.NewBuilder(buildToolIdentifier)
+	toc, err := ztocBuilder.BuildZtoc(layerTmpFile.Name(), ztocSpanSize, ztoc.WithCompression(compressionAlgo))
+	if err != nil {
+		return nil, err
+	}
+
+	ztocReader, ztocDesc, err := ztoc.Marshal(toc)
+	if err != nil {
+		return nil, err
+	}
+	ztocDigest := repb.Digest{
+		Hash:      ztocDesc.Digest.Encoded(),
+		SizeBytes: ztocDesc.Size,
 	}
-	return s.cache.Set(ctx, resourceName(digest), bytes)
+	ztocBytes, err := ioutil.ReadAll(ztocReader)
+	s.cache.Set(ctx, resourceName(&ztocDigest), ztocBytes)
```

Comment:
the cache interface has a Writer method that doesn't require providing the full buffer in one go.

Reply:
Good call, done